src/search/plugins/types.cc

```diff
@@ -0,0 +1,338 @@
+    : pointer_type(pointer_type), type_name(type_name), synopsis(synopsis),
+      predefinable(predefinable) {
+}
+
+bool FeatureType::operator==(const Type &other) const {
+    const FeatureType *other_ptr = dynamic_cast<const FeatureType *>(&other);
+    return other_ptr && pointer_type == other_ptr->pointer_type;
+}
+
+bool FeatureType::is_feature_type() const {
+    return true;
+}
+
+bool FeatureType::is_predefinable() const {
+    return predefinable;
+}
+
+string FeatureType::get_synopsis() const {
+    return synopsis;
+}
+
+string FeatureType::name() const {
+    return type_name;
+}
+
+size_t FeatureType::hash() const {
+    return std::hash<type_index>()(typeid(FeatureType)) ^ std::hash<type_index>()(pointer_type);
+}
+
+ListType::ListType(const Type &nested_type)
+    : nested_type(nested_type) {
+}
+
+bool ListType::operator==(const Type &other) const {
+    const ListType *other_ptr = dynamic_cast<const ListType *>(&other);
+    return other_ptr && nested_type == other_ptr->nested_type;
+}
+
+bool ListType::is_list_type() const {
+    return true;
+}
+
+bool ListType::has_nested_type() const {
+    return true;
+}
+
+const Type &ListType::get_nested_type() const {
+    return nested_type;
+}
+
+bool ListType::can_convert_into(const Type &other) const {
+    return other.is_list_type() && other.has_nested_type()
+           && nested_type.can_convert_into(other.get_nested_type());
+}
+
+string ListType::name() const {
+    return "list of " + nested_type.name();
+}
+
+size_t ListType::hash() const {
+    return std::hash<type_index>()(typeid(ListType)) ^ nested_type.hash();
+}
+
+bool EmptyListType::operator==(const Type &other) const {
+    const EmptyListType *other_ptr = dynamic_cast<const EmptyListType *>(&other);
+    return other_ptr;
+}
+
+bool EmptyListType::is_list_type() const {
+    return true;
+}
+
+bool EmptyListType::can_convert_into(const Type &other) const {
+    return other.is_list_type();
+}
+
+string EmptyListType::name() const {
+    return "empty list";
+}
+
+size_t EmptyListType::hash() const {
+    return std::hash<type_index>()(typeid(EmptyListType));
+}
+
+EnumType::EnumType(type_index type, const EnumInfo &documented_values)
+    : type(type), documented_values(documented_values) {
+    values.reserve(documented_values.size());
+    for (const auto &value_and_doc : documented_values) {
+        values.push_back(value_and_doc.first);
+    }
+}
+
+bool EnumType::operator==(const Type &other) const {
+    const EnumType *other_ptr = dynamic_cast<const EnumType *>(&other);
+    return other_ptr && type == other_ptr->type;
+}
+
+bool EnumType::is_enum_type() const {
+    return true;
+}
+
+int EnumType::get_enum_index(const string &value) const {
+    auto it = find(values.begin(), values.end(), value);
+    int enum_index = static_cast<int>(it - values.begin());
+    if (enum_index >= static_cast<int>(values.size()))
+        throw utils::Exception("Invalid enum value: " + value);
```

Comment:
Let E be an enum of size N. If the user inputs an invalid value, then enum_index=N which is an invalid index. We did not check this and this causes errors at random places. 

This fixes it, but which Exception should we throw?

Reply:
We handled this with a construct context